README

This directory contains "p2c" version 1.18, a Pascal to C translator.

"p2c"  Copyright 1989  Dave Gillespie
                       256-80 Caltech
                       Pasadena CA 91125
                       daveg@csvax.caltech.edu, cit-vax!daveg


This program is distributed under the terms of the GNU License Agreement.
See the file src/COPYING for details.

The GNU License Agreement restrictions do _not_ apply to code generated
by p2c, nor to the p2c run-time files "p2clib.c" and "p2c.h".



QUICK INSTRUCTIONS

The top-level Makefile in this directory tree knows how to build
p2c and run it on some example programs.  The compiled p2c will be
"installed" in this directory tree rather than in public directories.
(For a full public installation, see the instructions below.)
Just type "make test" to build p2c and run the examples.



FULL INSTALLATION INSTRUCTIONS

For a complete system, you must:

   a) Create a home directory for p2c, such as /usr/lib/p2c.  The exact
      location does not matter, as long as p2c knows how to find it.

   b) Create an include directory, such as /usr/include/p2c.  If you are
      unable to arrange for an include directory, you can edit sys.p2crc
      so as not to require one, or compile translated code with a
      suitable -I flag.

   c) Install the executable "p2c" in /usr/bin or elsewhere on the
      standard search path.

   d) Install the run-time library "libp2c.a" in /usr/lib or elsewhere
      on the standard library search path.

   e) Install the Unix manual entry "p2c.1" in /usr/man/man1.  If you
      have a "man1.Z" directory instead (for compressed man pages)
      you will have to edit the Makefile to use a "compress" command.

   f) Install the necessary files in the p2c home directory.

The Makefile is initially set up to create a private version of p2c in
the p2c distribution directory.  Just "cd src" and "make install" to
perform the above steps.  This will place the p2c executable, run-time
library, include directory, and man page in the "home" subdirectory of
the distribution directory.  The man page will be in human-readable form.

To compile code that has been translated by this private version of p2c,
use the command:

    cc -I. foo.c libp2c.a

(assuming "home," i.e., HOMEDIR, is the current directory.)

The "examples" subdirectory contains a few sample Pascal programs and
a Makefile to translate and compile them.

To install p2c for public use, edit the Makefile to indicate the correct
directory names.  Suggested names are included as comments.  Also change
MANFILE as shown to store the man page in the proper form for the "man"
command.

    HOMEDIR      default /usr/lib/p2c       Home directory
    INCDIR       default /usr/include/p2c   Include directory
    BINDIR       default /usr/bin           Directory for "p2c" executable
    LIBDIR       default /usr/lib           Directory for "libp2c.a" library
    MANDIR       default /usr/man/man1      Directory for "p2c.1" manual page

Make sure the home and include directories have been created and are
writable, and that the bin, lib, and man directories are writable.
(You may have to be root to do this.)  Recompile "p2c" with another
"make install".  Be sure "trans.c" is recompiled with the new value
of HOMEDIR; you can ensure this by doing "make newhome" before
"make install".  You can always type "make" with no arguments to
compile without installing.

HP Pascal Workstation System users will want to copy all of ../HP/include
into the new p2c include directory, and copy all of ../HP/import into
the p2c home directory.  You will have to write emulations for the
functions you use out of these modules.  (The HP directories are
present in the ftp distribution only---they have been omitted from
the shar'd version to save space.)

Incomplete interface texts for standard Turbo units are included in
turbo.imp.  Turbo users will have to flesh these out and write emulations
for whatever parts of the Turbo runtime library they need.

If you wish to edit various system-wide configuration parameters, it is
better to put them in loc.p2crc than in sys.p2crc.  As an example, on a
Turbo-oriented environment you may wish to add

    Language Turbo

to the loc.p2crc file to replace the normal default (HP Pascal).



PORTABILITY

P2c was originally developed on a homebrew C compiler on the HP Pascal
Workstation operating system.  Current development takes place on HP-UX,
which is System V based.  The translator has also been compiled and used
on Sun-3's, so presumably it is fairly portable among 32-bit Unix machines.

Many parts of the code will have to be adjusted in order for p2c to work
on a machine with 16-bit ints.  I have never had reason or opportunity to
track these down since I don't have access to such a machine.  Please note
that code _generated_ by p2c works with either size of int even though p2c
itself does not.

If your compiler supports ANSI prototypes, most of the 16-bit portability
problems will go away except for possible %d/%ld sloppiness in sprintf control
strings.  There may be places where p2c tries to cast a pointer into an
int, but I believe all such cases use longs consistently.  You will have
a problem if your long type is not large enough to store a pointer.

Because p2c was developed to run under Unix, it is sloppy about memory
allocation.  Some data structures are allocated but never freed.  Profiles
show that the waste is not too great in most cases, but if it is a problem
you may have to write a garbage collector (this shouldn't be too hard if you
only run the collector after translating each procedure).



src/HISTORY


Summary of changes to "p2c"
------- -- ------- --  ---


Version 1.19:

 * P2c's copyright has been assigned to the Free Software Foundation.

 * Improved p2c to preserve Pascal type names in most cases.

 * Introduced "PreserveTypes" (PreserveTypes=0 makes p2c work like v1.18).

 * Also added "PreservePointers" (default off) and "PreserveStrings" (on).

 * Changed p2c.h to recognize funny ANSI-like header files in HP-UX 7.0.

 * Changed to handle taking the address of an argument of a promotable type.

 * Changed parser to accept "a * -b" without syntax errors.

 * Fixed a few unreachable statement warnings in the p2c sources.

 * Changed exit(0) to exit(EXIT_SUCCESS) in main program.

 * Added support for VAX C and MSDOS to p2c.h.

 * Bug fix in parse.c:fixblock: Call filebasetype in similartypes arg.
   Symptom: Pascal GET procedure sometimes crashed p2c when StructFiles=1.

 * Bug fix in pexpr.c:p_funcarglist: Handle stdin/stdout when StructFiles=1.
   Symptom: myproc(output) generated "type mismatch" warning and bad code.

 * Bug fix in expr.c:makeexpr_rel: Use ISCONST to guard checkconst call.
   Symptom: "J <= A-I-1+C" where "const C=1" put p2c into an infinite loop.

 * Bug fix in decl.c:setupmeaning: Consider enum consts to be global names.
   Symptom: Local enum consts can conflict with similar global names.

 * Bug fix in lex.c:push_input/pop_input: Save/restore curtokbuf/curtokcase.
   Symptom: Value lost when main calls defmacro after setup_lex, before parser.

 * Bug fix in parse.c:out_function: Don't recursively output "exported" funcs.
   Symptom: This occurs when an "external" decl occurs local to a function.

 * Bug fix in p2c.h:AGETFBUF: Removed extra "&".
   Symptom: Taking address of an array generated a warning on some compilers.

 * Bug fix in p2c.h:LONG_MAX, etc.: Fixed to cast back to signed types.
   Symptom: LONG_MAX had type unsigned long; "(double)(-LONG_MAX)" failed.

 * Bug fix in p2clib.c:P_trimname: Added trailing null to output string.
   Symptom: Function wouldn't always work after first call.

 * Added to sys.p2crc/trans.h the following configuration parameter:

    FixPromotedArgs    Fix "&" of arguments which may be promoted.
    PreserveTypes      Use typedef to preserve most Pascal type names.
    PreservePointers   Use typedef specifically for pointer types.
    PreserveStrings    Use typedef specifically for string types.
    SpaceCommas	       Use spaces after commas (default on).
    SpaceFuncs	       Use spaces after function names (default off).
    NoBanner	       Suppress "From input file..." comment at top of file.

Version 1.18:

 * Improved indentation strategy for EMBED comments.

 * Allowed one-line EMBED comments: {EMBED foo();}

 * Made {SKIP} ... {NOSKIP} comments nestable.  ({NOSKIP ... } isn't nestable.)

 * Changed to emit comments in export sections of modules in correct order.

 * Rewrote decl.c:p_arraydecl slightly to avoid a gcc bug.

 * Bug fix in expr.c:makeexpr_sizeof: Compute size by hand if SkipIndices used.
   Symptom: Size was "sizeof(arr)" even if it should have omitted skipped part.

 * Bug fix in trans.h: Added #undef _toupper/_tolower at end.
   Symptom: Failed to compile on Apollos.

 * Bug fix in parse.c:fixblock: Added "sp->stm1 &&" check for WHILE statements.
   Symptom: "while cond do ;" caused a reference-to-NULL error.

 * Bug fix in decl.c:makefakestruct: Suppressed fakestruct when importing.
   Symptom: Importing a module generates two declarations for _REC_A.

 * Bug fix in Makefile: Added rules to make home dirs if not found.

Version 1.17:

 * Introduced "StructFile" stuff to help with more intensive file operations.

 * Simplified "N + 0.0" to "(double)N" instead of "N" when appropriate.

 * Fixed typo in sys.p2crc: "LogBreakPenalty" => "RelBreakPenalty".

 * Changed to avoid generating bit fields as members of a union.

 * Now "stringarray[i] := expr" will sprintf expr into lhs with no temp var.

 * Handled assignment of small packed arrays of chars into big ones.

 * Added to sys.p2crc/trans.h the following configuration parameters:

    CharConsts      Assume "const c='x'" is a char, not string constant.
    SelfIncludeName Format to use when a module includes its own .h file.
    StructFiles     Translate files as special structs rather than FILE *'s.
    StructFile      A file variable to be treated as if StructFiles=1.

Version 1.16:

 * Reduced wasteful proliferation of identical-looking temporary variables.

 * Improved heuristic for determining allocation size for set constructors.

 * Corrected handling for EOF applied to buffered binary files.

 * Bug fix in decl.c: Handled a rare case involving forward-declared pointers.

 * Bug fix in p2clib.c:P_remset: Forgot to fix s[0] if highest bit was cleared.

 * Bug fix in p2clib.c:P_setxor: Forgot to fix s[0] if highest bits cancelled.

 * Added to sys.p2crc/trans.h the following configuration parameters:

    EofBufName      Macro to use for EOF of a bufffered file.
    FilePosBufName  Macro to use for FILEPOS or POSITION of a buffered file.

Version 1.15:

 * Taught the line breaker to handle logical/relational operators specially.

 * Improved code generation for large set constructors.

 * Changed makeexpr_assign to handle EK_LITCAST's on the lefthand side.

 * Changed makeexpr_addr to handle casts.

 * Changed FieldMacros so that the substituted base expr retains its type.

 * Bug fix in p2c.h: Rewrote P_getbits_US and other small-packed-array macros.
   Symptom: On many architectures, these macros didn't work at all!
   NOTE: These now pack small arrays in the opposite direction (LSB first).

 * Bug fix in funcs.c:handleread_text: changed strread of string to use %255s.
   Symptom: Previous translation of %255c did not append a null character.

 * Bug fix in p2clib.c:strrtrim: Added test for empty string input.
   Symptom: Incorrect behavior on strrtrim("").

 * Bug fix in p2clib.c:P_addsetr: Changed to signed comparison of v1 and v2.

src/dir.c
/* "p2c", a Pascal to C translator.
   Copyright (C) 1989 David Gillespie.
   Author's address: daveg@csvax.caltech.edu; 256-80 Caltech/Pasadena CA 91125.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation (any version).

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA. */



#define define_parameters
#define PROTO_DIR_C
#include "trans.h"


/* This file is user-modifiable.  It is the "directory" of C functions
   for compiling in-line various Pascal library routines. */



extern void setup_module_hp();
extern void setup_module_cit();
extern void setup_module_tanner();




/* This function is called once when p2c is starting up, before
   the p2crc file has been read.
*/

void init_dir()
{


}





/* This function is called once when p2c is starting up, after
   the p2crc file has been read.
*/

void setup_dir()
{


}





/* This procedure is called after reading the import text for a module,
   where "name" is the module name, in upper-case letters.  Calls to
   "addmeaning", "makestandardfunc", etc. will annotate the context of
   the module.  Note that this will be called if the module is searched,
   even if it is never actually imported.
*/

#if 0
Static void _setup(name, defn)
char *name;
int defn;
{
    /* this is a dummy procedure which may be called by setup_module */
}
#endif

#define _setup(a,b)

void setup_module(name, defn)
char *name;
int defn;
{
    if (!strcicmp(name, "SYSTEM"))
	decl_builtins();
#ifdef CUST1
    CUST1(name, defn);
#endif
#ifdef CUST2
    CUST2(name, defn);
#endif
#ifdef CUST3
    CUST3(name, defn);
#endif
#ifdef CUST4
    CUST4(name, defn);
#endif
#ifdef CUST5
    CUST5(name, defn);
#endif
}





/* This procedure is called once after the p2crc file has been
   read and the built-in parameters have been "fixed".  It should
   check ranges and add defaults for any newly introduced parameters
   in the "rctable" (see "trans.h").
*/

void fix_parameters()
{


}





/* This function is called during a traversal of the tree of statements for
   a procedure.  Ordinarily it returns its argument; it may instead return
   an arbitrary other statement or sequence of statements, which will then
   be spliced in to replace the original one.  It may return NULL to delete
   the statement altogether.
*/

Stmt *fix_statement(sp)
Stmt *sp;
{
    return sp;
}





/* This is the analogous function for expression traversals.  It is
   called after the arguments have been (recursively) fixed and all
   built-in fixes have been performed.
*/

Expr *fix_expression(ex, env)
Expr *ex;
int env;
{
    return ex;
}





/* This procedure is called when fixing an expression of type
   EK_BICALL.  It is called before the arguments are fixed.  If
   it recognizes the BICALL, it should fix the arguments, then
   return a (possibly modified) fixed expression, which may or
   may not be a BICALL.  That expression will then be sent to
   fix_expression() as usual, but other standard fixes will not
   automatically be performed on it.  If the BICALL is not
   recognized, the function should return NULL.
*/

Expr *fix_bicall(ex, env)
Expr *ex;
int env;
{
    return NULL;
}





/* This function returns nonzero if the built-in function "name"
   should be written "if (f(x))" rather than "if (f(x) != 0)"
   when used as a boolean.  The call does *not* necessarily have
   to return a 1-or-0 value.
*/

int boolean_bicall(name)
char *name;
{
    return (!strcmp(name, "strcmp") ||
            !strcmp(name, "strncmp") ||
            !strcmp(name, "memcmp") ||
            !strcmp(name, "feof") ||
            !strcmp(name, "feoln"));
}





/* The function "name" promises not to change certain of its
   VAR-style parameters.  For each of arguments i = 0 through 15,
   if bit 1<<i of the return value of this function is set, and
   the i'th parameter is a pointer to an object, then the function
   guarantees not to change that object.
*/

unsigned int safemask_bicall(name)
char *name;
{
    Symbol *sp;

    sp = findsymbol_opt(name);
    if (sp) {
	if (sp->flags & (STRUCTF|STRLAPF))
	    return ~1;
	if (sp->flags & (NOSIDEEFF|DETERMF))
	    return ~0;
    }
    if (!strcmp(name, "fwrite") ||
        !strcmp(name, "memchr"))
        return 1;
    if (!strcmp(name, "memcpy") ||
        !strcmp(name, "memmove"))
        return 2;
    if (!strcmp(name, "memcmp"))
        return 3;
    if (!strcmp(name, "sprintf") ||
        !strcmp(name, "fprintf"))
        return ~1;
    if (!strcmp(name, "printf"))
        return ~0;
    return 0;
}





/* The function "name" has side effects that could affect other variables
   in the program besides those that are explicitly mentioned.
*/

int sideeffects_bicall(name)
char *name;
{
    return 0;
}






/* End. */



